Testnet/NFTAuctionStore-Royalties/NFTAuctionStore/NFTAuctionStore.cs

```diff
@@ -0,0 +1,287 @@
+
+            Assert(result.Success, "Transfer failed.");
+        }
+
+        if (royaltyAmount > 0)
+        {
+            var royaltyTransfer = Transfer(royaltyRecipient, royaltyAmount);
+
+            Assert(royaltyTransfer.Success, "Royalty transfer failed.");
+            Log(new RoyaltyPaidLog { Recipient = royaltyRecipient, Amount = royaltyAmount });
+        }
+
+        TransferToken(contract, tokenId, Address, auction.HighestBidder);
+
+        Log(new AuctionEndSucceedLog { Contract = contract, TokenId = tokenId, HighestBidder = auction.HighestBidder, HighestBid = auction.HighestBid });
+    }
+
+    public bool OnNonFungibleTokenReceived(Address operatorAddress, Address fromAddress, UInt256 tokenId, byte[] data)
+    {
+        EnsureNotPayable();
+
+        var tokenContract = Message.Sender;
+
+        Assert(State.IsContract(tokenContract), "The Caller is not a contract.");
+
+        var seller = fromAddress == Address.Zero ? operatorAddress : fromAddress;
+
+        var parameters = Serializer.ToStruct<AuctionParam>(data);
+
+        Assert(parameters.StartingPrice > 0, "Price should be higher than zero.");
+
+        Assert(Address == GetOwner(tokenContract, tokenId), "The store contract is not owner of the token.");
+
+        var auction = GetAuctionInfo(tokenContract, tokenId);
+
+        Assert(!auction.OnAuction, "The token is already on auction.");
+
+        auction = new AuctionInfo
+        {
+            Seller = seller,
+            EndBlock = checked(Block.Number + parameters.Duration),
+            StartingPrice = parameters.StartingPrice,
+            OnAuction = true
+        };
+
+        SetAuctionInfo(tokenContract, tokenId, auction);
+
+        Log(new AuctionStartedLog { Contract = tokenContract, TokenId = tokenId, EndBlock = auction.EndBlock, Seller = auction.Seller, startingPrice = auction.StartingPrice });
+
+        return true;
+    }
+
+    private void TransferToken(Address contract, UInt256 tokenId, Address from, Address to)
+    {
+        var result = Call(contract, 0, "TransferFrom", new object[] { from, to, tokenId });
+
+        Assert(result.Success, "The token transfer failed.");
+    }
+
+
+    private Address GetOwner(Address contract, UInt256 tokenId)
+    {
+        var result = Call(contract, 0, "OwnerOf", new object[] { tokenId });
+
+        Assert(result.Success && result.ReturnValue is Address, "GetOwner method call failed.");
+
+        return (Address)result.ReturnValue;
+    }
+
+    private void EnsureNotPayable() => Assert(Message.Value == 0, "The method is not payable.");
+
+    private bool SupportsRoyaltyInfo(Address contract)
+    {
+        // 6 is IRoyaltyInfo
+        var supportsInterfaceResult = Call(contract, 0, "SupportsInterface", new object[] { 6 });
+
+        return supportsInterfaceResult.Success && ((bool)supportsInterfaceResult.ReturnValue);
+    }
+
+    private object[] GetRoyaltyInfo(Address contract, UInt256 tokenId, ulong salePrice)
```

Comment:
> I still didn't understand why we introduce new contracts instead of override current ones. We overrided ERC-721 implementation multiple times already.
> 
> Should we avoid overriding in the case of fixing a bug on the contract ?

Having a new contract feels logical because each contract maps to whitelisted bytecode. If we update an existing contract then the old code is buried.